### PR TITLE
0.6.x

### PR DESCRIPTION
**Pull request by nexeh**
I have created a chart gimmick that reads data from an HTML table that resulted from paring markdown. This solution uses Chart.js to render the Chart in a canvas. Licensing is MIT. The gimmick does use a 100 milli timeout that i think you should look at.

This pull is in reference to issue #157.

### CODE REVIEW

**Comment by Dynalon** [comment]
I haven't forgotten you, i am a weekend commuter and for some reasons currently offline after work during weekdays, I'll take a look at it on the weekend.

File: js/gimmicks/chart.js [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+        name: 'chart',
+        version: $.md.version,
+        once: function() {
+            $.md.linkGimmick(this, 'chart', chart);
+
+            // load the chart js
+            $.md.registerScript(this, scripturl, {
+                license: 'EXCEPTION',
+                loadstage: 'skel_ready',
+                finishstage: 'bootstrap'
```
Comment (Dynalon) [inline js/gimmicks/chart.js]: Change the finishstage from 'boostrap' to 'gimmick' - That way the replacement of the links has finished at the 'gimmick' stage - (we will use that later)

File: js/gimmicks/chart.js [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+                                chartConfig.datasets[i].data.push(this.textContent);
+                            }
+                        }
+                    }
+                });
+            });
+
+            setTimeout(function(){
+                new Chart(document.getElementById(options.canvasId).getContext("2d"))[options.chartType](chartConfig, options.chartOptions);
+            }, 100);
```
Comment (Dynalon) [inline js/gimmicks/chart.js]: We can get rid of the setTimeout by using the (sadly undocumented) way of MDwiki to subscribe to the "execution phases". We subscribe into the "postgimmick" phase, which will called after the 'gimmick' phase that I mentioned earlier. Example code:

```js
$.md.stages('postgimmick').subscribe(function() {
    new Chart(document.getElementById(options.canvasId).getContext("2d"))[options.chartType](chartConfig, options.chartOptions);
});
```

**Comment by Dynalon** [comment]
I added some inline comments, these are just minor additions and you couldn't possibly know of them because I sadly didn't document this gimmick stuff/API anywhere. I made the changes on my machien and it works great, this is the best and most complex gimmick contributio so far!

If you make the changes and push to your branch I can merge the pull request (provided you signed the CLA, see http://dynalon.github.io/mdwiki/#!contribute/index.md).

I got another suggestion that I'd would ask you to do, but if you don't have time it is ok, I will merge it anyways:

Using the jQuery("th") selector and iterating over all the tables in the document seems awkward and has performance implications. I'd rather see something like selecting the parent <table> tag once and only operate on it's childs. For example:

```js
// somehow find what table colum/cell <th> this chart is made for
var column = findMyColumn();
// select the immediate <table> parent - there should only be one because we dont allow nested tables
var $tableRoot = $(column).parents('table');

// now whenever we need to access an element that is within the table:
var allHeadings = $tableRoot.find('th');
var allRows = $tableRoot.find("td");
```

Within the `.find()` you can pass any jQuery selector that would else have been passed into the jQuery() object - but I think you can remove lots of the for-iterations then to reduce code complexity. Subselects are of course also possible: `$tableRoot.find('td').find('a').find('img');`  etc.

File: js/gimmicks/chart.js [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+
+    var chartGimmick = {
+        name: 'chart',
+        version: $.md.version,
+        once: function() {
+            $.md.linkGimmick(this, 'chart', chart);
+
+            // load the chart js
+            $.md.registerScript(this, scripturl, {
+                license: 'EXCEPTION',
```
Comment (Dynalon) [inline js/gimmicks/chart.js]: As to my finding, chartsJs is licensed under MIT license. So we should change 'EXCEPTION' to 'MIT' (MIT is compatible with MDwikis GPL/AGPL).

**Comment by nexeh** [comment]
Thanks for the review! I will make those changes ASAP. Today is my Sprint end so i have a feeling it will be an all hands on deck kind of day but ill try to squeeze it in! If you don't mind i will do the optimizations to table processing as a separate pull request. I can definitely clean that area up. I was in the mind set of 'lets make it work, see what he thinks, and then optimize'. As you can imagine, i wasn't sure what you would think about this new gimmick and i didn't want to waste to much time to find out you didn't like it!

**Comment by Dynalon** [comment]
I actually gave you a code example with a typo, it is `$.md.stage('postgimmick')`   and forgot that you need to call done(). For some reason it didn't work anymore on my machine. I've now combined the $.md.stage stuff with a setTimeout of 0 seconds - this will just put the function to the end of execution queue. I have no idea whats the issue for now, need to dig deeper. but for now it is ok. Have a look here to see the changes: https://github.com/Dynalon/mdwiki/blob/0.6.x/js/gimmicks/chart.js#L104

**Comment by nexeh** [comment]
Woah, you are right. this is odd because i tested your changes too and it worked. I just got back from lunch, removed my current changes for another enhancement i was exploring, and it no longer works! How strange

**Comment by nexeh** [comment]
I pulled your changes into my fork, tested, and everything looks good.

**Pull request merged**